cuML/src/solver/cd.h

```diff
@@ -0,0 +1,218 @@
+
+	allocate(loss_value, 1);
+	allocate(pred, n_rows, true);
+	allocate(residual, n_rows, true);
+	allocate(squared, n_cols, true);
+
+	std::vector<math_t> h_coef(n_cols, math_t(0));
+
+	if (fit_intercept) {
+		allocate(mu_input, n_cols);
+		allocate(mu_labels, 1);
+		if (normalize) {
+			allocate(norm2_input, n_cols);
+		}
+
+		GLM::preProcessData(input, n_rows, n_cols, labels, intercept, mu_input,
+				mu_labels, norm2_input, fit_intercept, normalize, cublas_handle,
+				cusolver_handle, stream);
+	}
+
+	std::vector<int> ri(n_cols);
+	std::mt19937 g(rand());
+	initShuffle(ri, g);
+
+	math_t l2_alpha = (1 - l1_ratio) * alpha * n_rows;
+	alpha = l1_ratio * alpha * n_rows;
+
+	if (normalize) {
+		math_t scalar = math_t(1.0) + l2_alpha;
+		Matrix::setValue(squared, squared, scalar, n_cols, stream);
+	} else {
+		LinAlg::colNorm(squared, input, n_cols, n_rows, LinAlg::L2Norm, false,
+				stream);
+		LinAlg::addScalar(squared, squared, l2_alpha, n_cols, stream);
+	}
+
+	copy(residual, labels, n_rows);
+
+	for (int i = 0; i < epochs; i++) {
+		if (i > 0 && shuffle) {
+			Solver::shuffle(ri, g);
+		}
+
+		math_t coef_max = 0.0;
+		math_t d_coef_max = 0.0;
+		math_t coef_prev = 0.0;
+
+		for (int j = 0; j < n_cols; j++) {
+			int ci = ri[j];
+			math_t *coef_loc = coef + ci;
+			math_t *squared_loc = squared + ci;
+			math_t *input_col_loc = input + (ci * n_rows);
+
+			LinAlg::multiplyScalar(pred, input_col_loc, h_coef[ci], n_rows, stream);
+			LinAlg::add(residual, residual, pred, n_rows, stream);
+			LinAlg::gemm(input_col_loc, n_rows, 1, residual, coef_loc, 1, 1,
+								CUBLAS_OP_T, CUBLAS_OP_N, cublas_handle, stream);
+
+			if (l1_ratio > math_t(0.0))
+				Functions::softThres(coef_loc, coef_loc, alpha, 1, stream);
+
+			LinAlg::eltwiseDivideCheckZero(coef_loc, coef_loc, squared_loc, 1, stream);
+
+			coef_prev = h_coef[ci];
+			updateHost(&(h_coef[ci]), coef_loc, 1);
+			math_t diff = abs(coef_prev - h_coef[ci]);
+
+			if (diff > d_coef_max)
+				d_coef_max = diff;
+
+			if (abs(h_coef[ci]) > coef_max)
+				coef_max = abs(h_coef[ci]);
+
+			LinAlg::multiplyScalar(pred, input_col_loc, h_coef[ci], n_rows, stream);
+			LinAlg::subtract(residual, residual, pred, n_rows, stream);
+		}
+
+		bool flag_continue = true;
+		if (coef_max == math_t(0)) {
+			flag_continue = false;
+		}
+
+		if ((d_coef_max / coef_max) < tol) {
+			flag_continue = false;
+		}
+
+		if (!flag_continue) {
+			//printf("iter:%d, coef_max: %f, d_coef_max: %f\n", i, coef_max, d_coef_max);
+			break;
+		}
+	}
+
+	if (fit_intercept) {
+		GLM::postProcessData(input, n_rows, n_cols, labels, coef, intercept,
+				mu_input, mu_labels, norm2_input, fit_intercept, normalize,
+				cublas_handle, cusolver_handle, stream);
+
+		if (mu_input != NULL)
+			CUDA_CHECK(cudaFree(mu_input));
+		if (mu_labels != NULL)
+			CUDA_CHECK(cudaFree(mu_labels));
+		if (normalize) {
+			if (norm2_input != NULL)
+				cudaFree(norm2_input);
+		}
+	} else {
+		*intercept = math_t(0);
+	}
+
+	if (pred != NULL)
```

Comment:
@teju85, I recently noticed a preference towards nullptr rather than NULL in Chirayu's PR. Is this something else we should be striving to make consistent?

Reply:
Yes, since we are atleast c++11, using `nullptr` is cleaner.